core/shared/src/main/scala/scalaz/zio/Duration.scala

```diff
@@ -0,0 +1,25 @@
+package scalaz.zio
+
+sealed trait Duration {
+  def toDuration: scala.concurrent.duration.Duration
+}
+
+object Duration {
+
+  case class Finite(duration: scala.concurrent.duration.FiniteDuration) extends Duration {
+    final def toDuration = duration
+  }
+
+  case class Infinite() extends Duration {
+    final def toDuration = scala.concurrent.duration.Duration.Inf
+  }
+
+  def fromNanos(length: Long): Finite = Finite(scala.concurrent.duration.Duration.fromNanos(length))
+
+  def fromDuration(duration: scala.concurrent.duration.Duration): Either[IllegalArgumentException, Duration] =
+    duration match {
+      case finiteDuration: scala.concurrent.duration.FiniteDuration => Right(Finite(finiteDuration))
+      case inf if inf == scala.concurrent.duration.Duration.Inf     => Right(Infinite())
+      case x                                                        => Left(new IllegalArgumentException(s"Woah dude, $x is not a meaningful zio.Duration!"))
```

Comment:
:laughing: sorry, this was just a sketch, maybe we should come up with a better message

Reply:
I'm not convinced an exception adds value. I'd suggest to change to `Option[Duration]`.

Reply:
> not convinced an exception adds value

hmm, I think it does -- it has a stack trace, so if you have a very big and complex monadic `Either` program, you can still see precisely where the error occurred, if you somehow endup with `Left(exn)`
exceptions are still useful, we just shouldn't be `throw`ing them :)

Reply:
Well let's change to `Option` and the user can add their own custom exception (maybe of some typed error). It's very opinionated to choose a particular exception for this.